### PR TITLE
Add feature to remove sudo

### PR DESCRIPTION
**Pull request by sea212**
This PR adds the `without-sudo` feature to remove the sudo pallet from runtime and any initialization from genesis.
I would like to keep sudo as a default feature, because that way anybody can build a node from source without any arguments and have sudo in the standalone chain available.
I decided to use a `without-sudo` feature instead of `with-sudo`, because by default sudo should be enabled. One approach would be to add `with-sudo` to the `default` feature, but in that case disabling it would require disabling all other default features as well afaik `--no-default-features`, which seemed not ideal to me.

### CODE REVIEW

**Review by vivekvpandya — APPROVED** [review]
Looks good to me! Is there any relevant bug to understand the context?

File: runtime/Cargo.toml [hunk 1]
```diff
@@ -20,7 +20,7 @@ pallet-preimage = { branch = "moonbeam-polkadot-v0.9.19", default-features = fal
 pallet-proxy = { branch = "moonbeam-polkadot-v0.9.19", default-features = false, git = "https://github.com/purestake/substrate" }
 pallet-randomness-collective-flip = { branch = "moonbeam-polkadot-v0.9.19", default-features = false, git = "https://github.com/purestake/substrate" }
 pallet-scheduler = { branch = "moonbeam-polkadot-v0.9.19", default-features = false, git = "https://github.com/purestake/substrate" }
-pallet-sudo = { branch = "moonbeam-polkadot-v0.9.19", default-features = false, git = "https://github.com/purestake/substrate" }
+pallet-sudo = { branch = "moonbeam-polkadot-v0.9.19", default-features = false, optional = true, git = "https://github.com/purestake/substrate" }
```
Comment (Chralt98) [inline runtime/Cargo.toml:23]: Because of this, shall we add a `?` optional [here](https://github.com/zeitgeistpm/zeitgeist/blob/f3db56ff8945ffef19e696bc8d6c026b753c2a39/runtime/Cargo.toml#L306)? According to https://doc.rust-lang.org/cargo/reference/features.html#dependency-features

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -221,35 +220,54 @@ macro_rules! create_zeitgeist_runtime {
...
+}
+
+#[cfg(not(feature = "without-sudo"))]
+create_zeitgeist_runtime_with_additional_pallets!(
+    // Others
+    Sudo: pallet_sudo::{Call, Config<T>, Event<T>, Pallet, Storage} = 150,
 );
 
+#[cfg(feature = "without-sudo")]
+create_zeitgeist_runtime_with_additional_pallets!();
```
Comment (Chralt98) [inline runtime/src/lib.rs:269]: Why is this required here?

File: runtime/src/lib.rs [hunk 3]
```diff
@@ -364,12 +383,15 @@ cfg_if::cfg_if! {
                     | Call::Utility(_)
                     | Call::Vesting(_) => true,
 
+                    #[cfg(not(feature = "without-sudo"))]
+                    Call::Sudo(_) => true,
```
Comment (Chralt98) [inline runtime/src/lib.rs:387]: Maybe depend on the compiler flag check feature [here](https://github.com/zeitgeistpm/zeitgeist/blob/f3db56ff8945ffef19e696bc8d6c026b753c2a39/runtime/src/lib.rs#L419) too.

**Review by Chralt98 — CHANGES_REQUESTED** [review]
Ok, checked it. There are some possible flaws. But all in all good job to go in the direction of full decentralisation.

File: node/src/chain_spec/dev.rs [hunk 4]
```diff
@@ -54,6 +54,7 @@ pub fn dev_config(
                 .into_iter()
                 .map(|acc| EndowedAccountWithBalance(acc, INITIAL_BALANCE))
                 .collect(),
+                #[cfg(not(feature = "without-sudo"))]
```
[the remark this answers is not in the document]
Comment (Chralt98) [reply node/src/chain_spec/dev.rs:57]: Oh my fault, because of this
Bildschirmfoto 2022-07-08 um 09 18 26

File: node/src/chain_spec/mod.rs [hunk 5]
```diff
@@ -300,6 +303,7 @@ fn endowed_accounts_staging_mainnet() -> Vec<EndowedAccountWithBalance> {
             hex!["b449a256f73e59602eb742071a07e4d94aaae91e6872f28e161f34982a0bfc0d"].into(),
             DEFAULT_COLLATOR_BALANCE_MAINNET.unwrap(),
         ),
+        #[cfg(not(feature = "without-sudo"))]
         // dE2nxuZc5e7xBbU1cGikmtVGws9niNPUayigoDdyqB7hzHQ6X
```
[the remark this answers is not in the document]
Comment (sea212) [reply node/src/chain_spec/mod.rs:307]: What is the difference? Why prefer one over the other?

File: node/src/chain_spec/dev.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (sea212) [reply node/src/chain_spec/dev.rs:57]: This is not the case, the next line after the feature is `get_account_id_from_seed::<sr25519::Public>("Alice"),`, which defines `Alice` as the superuser account.

File: runtime/Cargo.toml (hunk 1 above)
Comment (sea212) [reply runtime/Cargo.toml:23]: Good point, this was the initial idea why I made it optional. In the first iteration I used `pallet-sudo` and then `with-sudo` features. Turns out one cannot use the inherent `pallet-sudo` feature with the conditional inclusion `?`, and when using `with-sudo` one can use the `?` operator, but since `with-sudo` would be the default option, disabling it would require the rather unintuitive and too powerful flag `--no-default-features`. When using `without-sudo`, the `?` can't be used because `pallet-sudo` is used by default and only disabled with the feature.
So the way to go is to remove the optional parameter.

```suggestion
pallet-sudo = { branch = "moonbeam-polkadot-v0.9.19", default-features = false, git = "https://github.com/purestake/substrate" }
```

File: runtime/src/lib.rs (hunk 2 above)
Comment (sea212) [reply runtime/src/lib.rs:269]: I found no way to include that feature requirement inside a macro, so I had to create a new wrapper macro and apply the feature requirement on the whole wrapper macro. Let me know if you know a better approach to this, I already wasted like 20-30 minutes figuring out a nicer approach to that.

File: runtime/Cargo.toml (hunk 1 above)
Comment (Chralt98) [reply runtime/Cargo.toml]: Thank you!

File: runtime/src/lib.rs (hunk 3 above)
Comment (sea212) [reply runtime/src/lib.rs:387]: We don't need it in the last case, because when sudo is included, it falls into the `_ => ` match arm and calls won't be filtered. When it's not included it is not regarded at all.

File: node/src/chain_spec/battery_station.rs [hunk 6]
```diff
@@ -20,6 +22,7 @@ pub fn battery_station_staging_config(
                     parachain_id,
                 ),
                 endowed_accounts_staging_testnet(),
+                #[cfg(not(feature = "without-sudo"))]
                 root_key_staging_testnet(),
                 zeitgeist_wasm,
             )
```
[the remark this answers is not in the document]
Comment (sea212) [reply node/src/chain_spec/battery_station.rs:28]: Yes, this is correct.

File: node/src/chain_spec/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (sea212) [reply node/src/chain_spec/mod.rs:307]: Technically it does not make a difference (yes, comment are ignored), but the consistency argument is solid :+1:

File: runtime/src/lib.rs (hunk 3 above)
Comment (Chralt98) [reply runtime/src/lib.rs:387]: I understand, you mean because of not including the whole `Sudo` pallet in the `construct_runtime`. That makes sense.

**Review by Chralt98 — APPROVED** [review]
LGTM. Let us go in the direction of full decentralisation together!

File: runtime/src/lib.rs (hunk 2 above)
Comment (Chralt98) [reply runtime/src/lib.rs:269]: Sorry, I am out here. Maybe you @maltekliemann or @vivekvpandya ?

Comment (vivekvpandya) [reply runtime/src/lib.rs:269]: I also don't have anything better than what @sea212  has already done https://stackoverflow.com/questions/63008781/check-a-feature-flag-in-an-item-defined-in-a-macro
